core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala

```diff
@@ -0,0 +1,171 @@
+                                                  log: Logging): Future[KubernetesContainer] = {
+    implicit val tid = transid
+
+    val podName = name.replace("_", "-").replaceAll("[()]", "").toLowerCase()
+
+    val environmentArgs = environment.flatMap {
+      case (key, value) => Seq("--env", s"$key=$value")
+    }.toSeq
+
+    val labelArgs = labels.map {
+      case (key, value) => s"$key=$value"
+    } match {
+      case Seq() => Seq()
+      case pairs => Seq("-l") ++ pairs
+    }
+
+    val args = Seq("--generator", "run-pod/v1", "--restart", "Always", "--limits", s"memory=${memory.toMB}Mi") ++ environmentArgs ++ labelArgs
+
+    for {
+      id <- kubernetes.run(podName, image, args).recoverWith {
+        case _ => Future.failed(WhiskContainerStartupError(Messages.resourceProvisionError))
+      }
+      ip <- kubernetes.inspectIPAddress(id).recoverWith {
+        // remove the container immediately if inspect failed as
+        // we cannot recover that case automatically
+        case _ =>
+          kubernetes.rm(id)
+          Future.failed(WhiskContainerStartupError(Messages.resourceProvisionError))
+      }
+    } yield new KubernetesContainer(id, ip)
+  }
+
+}
+
+/**
+ * Represents a container as run by kubernetes.
+ *
+ * This class contains OpenWhisk specific behavior and as such does not necessarily
+ * use kubernetes commands to achieve the effects needed.
+ *
+ * @constructor
+ * @param id the id of the container
+ * @param addr the ip & port of the container
+ */
+class KubernetesContainer(protected val id: ContainerId, protected val addr: ContainerAddress)(
+  implicit kubernetes: KubernetesApi,
+  protected val ec: ExecutionContext,
+  protected val logging: Logging)
+    extends Container {
+
+  /** The last read timestamp in the log file */
+  private val lastTimestamp = new AtomicReference[Option[LocalDateTime]](None)
+
+  protected val waitForLogs: FiniteDuration = 2.seconds
+
+  // no-op under Kubernetes
+  def suspend()(implicit transid: TransactionId): Future[Unit] = Future.successful({})
+
+  // no-op under Kubernetes
+  def resume()(implicit transid: TransactionId): Future[Unit] = Future.successful({})
+
+  override def destroy()(implicit transid: TransactionId): Future[Unit] = {
+    super.destroy()
+    kubernetes.rm(id)
+  }
+
+  def logs(limit: ByteSize, waitForSentinel: Boolean)(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    import KubernetesClient.K8SDateTimeFormat
+
+    val activationMarkerCheck: ByteString ⇒ Boolean = { bs ⇒
+      val str = bs.utf8String
+      val parsed = str.parseJson.convertTo[LogLine]
+      // If this line is a sentinel line, remember the timestamp for lastTimestamp
+      if (bs.containsSlice(DockerContainer.ActivationSentinel)) {
+        Try(LocalDateTime.parse(parsed.time, K8SDateTimeFormat)).foreach(time => lastTimestamp.set(Option(time)))
+        true
+      } else false
+    }
```

Comment:
Is it not feasible to alter the entire stream for the kube-case to avoid to reparse all the loglines just to set the latest timestamp?

You could, for instance, return a fully parsed object (`LogLine` or something like a new `TypedLogLine(Instant, Stream, String)`?) from `kubernetes.logs`, drop the Framing, adjust limitWeighted and then have a `.map(_.toJson.compactPrint)` as the very last stage to be pluggable into the logstores.

You're parsing everything anyway, why not keep it that way? Is it just test-parity?